dex/farm/src/lib.rs

```diff
@@ -613,4 +613,207 @@ pub trait Farm:
+            reward.token_identifier == reward_token_id,
+            b"bad reward token id"
+        );
+
+        // The actual work starts here
+        self.reward_reserve().update(|x| *x += &reward.amount);
+
+        let old_attrs: FarmTokenAttributes<Self::Api> = self
+            .blockchain()
+            .get_esdt_token_data(
+                &self.blockchain().get_sc_address(),
+                &old_position.token_identifier,
+                old_position.token_nonce,
+            )
+            .decode_attributes()
+            .unwrap();
+
+        // Do not call burn_farm_tokens since this farm tokens belong to other contract
+        // which already updated its farm token supply counter.
+        self.send().esdt_local_burn(
+            &old_position.token_identifier,
+            old_position.token_nonce,
+            &old_position.amount,
+        );
+
+        let new_pos_token_id = self.farm_token_id().get();
+        let new_pos_amount = old_position.amount;
+
+        // Use this function because it also updates the farm token supply for this contract instance.
+        let new_pos_nonce = self.mint_farm_tokens(&new_pos_token_id, &new_pos_amount, &old_attrs);
+
+        let orig_caller = orig_caller_opt
+            .into_option()
+            .unwrap_or_else(|| caller.clone());
+
+        // Use this function since it works regardless of wasm ocasional unalignment.
+        self.transfer_execute_custom(
+            &orig_caller,
+            &new_pos_token_id,
+            new_pos_nonce,
+            &new_pos_amount,
+            &OptionalArg::None,
+        );
+
+        EsdtTokenPayment::new(new_pos_token_id, new_pos_nonce, new_pos_amount)
+    }
+
+    // Each farm that will be migrated and the newer version to which we migrate to
+    // will have to be configured using this function.
+    #[only_owner]
+    #[endpoint(setFarmConfiguration)]
+    fn set_farm_configuration(
+        &self,
+        is_old: bool,
+        old_farm_address: ManagedAddress,
+        old_farm_token_id: TokenIdentifier,
+        new_farm_address: ManagedAddress,
+    ) {
+        self.farm_configuration().set(&FarmContractConfig {
+            is_old,
+            old_farm_address,
+            old_farm_token_id,
+            new_farm_address,
+        });
+    }
+
+    // We also need to get the rps and transfer it to the new SC.
+    #[only_owner]
+    #[endpoint(stopRewardsAndGetRps)]
+    fn stop_rewards_and_get_rps(&self) -> BigUint {
+        assert!(self, !self.farm_configuration().is_empty(), b"empty config");
+        let config = self.farm_configuration().get();
+        assert!(self, config.is_old, b"bad config");
+
+        self.end_produce_rewards();
+        self.reward_per_share().get()
+    }
+
+    // In the new sc, we have to set the rps, so rewards can continue
+    // with positions being untouched.
+    #[only_owner]
+    #[endpoint(setRpsAndStartRewards)]
+    fn set_rps_and_start_rewards(&self, rps: BigUint) {
```

Comment:
This is a nice IDEA! It is somewhat dangerous. My first thought was that we make the transition process as normal and users will not get rewards on the old contract - only from the new - and the new contract will start accumulating position as normal. But this is much beneficial for the user - as it would get the rewards, and there is no need to rush for the migration.